### PR TITLE
log dump when test failed

### PR DESCRIPTION
**Pull request by xiaojingchen**
@weekface @cwen0 @zyguan @shuijing198799 PTAL

### CODE REVIEW

File: tests/cmd/e2e/main.go [hunk 1]
```diff
@@ -53,9 +53,11 @@ func main() {
 		LogLevel:       "2",
 	}
 	if err := oa.CleanOperator(operatorInfo); err != nil {
+		oa.DumpAllLogs(operatorInfo, []*tests.TidbClusterInfo{})
```
Comment (weekface) [inline tests/cmd/e2e/main.go]:
```suggestion
		oa.DumpAllLogs(operatorInfo, nil)
```

File: tests/cmd/e2e/main.go [hunk 2]
```diff
@@ -53,9 +53,11 @@ func main() {
...
 		glog.Fatal(err)
 	}
 	if err := oa.DeployOperator(operatorInfo); err != nil {
+		oa.DumpAllLogs(operatorInfo, []*tests.TidbClusterInfo{})
```
Comment (weekface) [inline tests/cmd/e2e/main.go]:
```suggestion
		oa.DumpAllLogs(operatorInfo, nil)
```

File: tests/manifests/e2e.yaml [hunk 3]
```diff
@@ -27,7 +27,15 @@ spec:
   serviceAccount: tidb-operator-e2e
   containers:
   - name: tidb-operator-e2e
-    image: ""
+    image: "localhost:5000/pingcap/tidb-operator-e2e:latest"
```
Comment (weekface) [inline tests/manifests/e2e.yaml]:
```suggestion
    image: ""
```
?

File: tests/log_dump.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		return err
+	}
+	defer logFile.Close()
+	plogFile, err := os.Create(filepath.Join(logPath, fmt.Sprintf("%s-%s-p.log", pod.Name, pod.Namespace)))
+	if err != nil {
+		return err
+	}
+	defer plogFile.Close()
+	logWriter := bufio.NewWriter(logFile)
+	plogWriter := bufio.NewWriter(plogFile)
```
Comment (zyguan) [inline tests/log_dump.go]:
```go
defer logWriter.Flush()
defer plogWriter.Flush()
```

File: tests/log_dump.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		dumpLog(fmt.Sprintf("kubectl logs -n %s %s -c %s", pod.Namespace, pod.GetName(), c.Name), logWriter)
+		dumpLog(fmt.Sprintf("kubectl logs -n %s %s -c %s -p", pod.Namespace, pod.GetName(), c.Name), plogWriter)
+	}
+
+	return nil
+}
+
+func dumpLog(cmdStr string, writer *bufio.Writer) {
+	writer.WriteString(fmt.Sprintf("$ %s\n", cmdStr))
+	data, err := exec.Command("/bin/sh", "-c", "/usr/local/bin/"+cmdStr).CombinedOutput()
```
Comment (zyguan) [inline tests/log_dump.go]:
```go
cmd := exec.Command(...)
cmd.Stdout, cmd.Stderr = writer, writer
err := cmd.Run()
```

File: tests/log_dump.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	if err != nil {
+		writer.WriteString(err.Error())
+	}
+	writer.WriteString(string(data))
```
Comment (zyguan) [inline tests/log_dump.go]: `data` might be nil, however it's unnecessary if you set writer as stdout & stderr.

File: tests/log_dump.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	if err != nil {
+		return err
+	}
+	defer resourceLogFile.Close()
+	resourceWriter := bufio.NewWriter(resourceLogFile)
+	dumpLog("kubectl get po -owide -n kube-system", resourceWriter)
+	dumpLog(fmt.Sprintf("kubectl get po -owide -n %s", operatorInfo.Namespace), resourceWriter)
+	dumpLog("kubectl get pv", resourceWriter)
+	dumpLog("kubectl get pv -oyaml", resourceWriter)
+	for _, testCluster := range testClusters {
```
Comment (zyguan) [inline tests/log_dump.go]: If two clusters have same namespace, we dump them twice to same log files.

File: tests/log_dump.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+			}
+			dumpedNamespace[testCluster.Namespace] = true
+		}
+	}
+
+	return nil
+}
+
+func dumpPod(logPath string, pod *corev1.Pod) error {
+	logFile, err := os.Create(filepath.Join(logPath, fmt.Sprintf("%s-%s.log", pod.Name, pod.Namespace)))
```
Comment (shuijing198799) [inline tests/log_dump.go]: log file name generate rule may resulting in the log generated during the second run will overwrite the first run

**Comment by weekface** [comment]
@xiaojingchen Please resolve the conflicts.

**Pull request merged**